### PR TITLE
Make no-autoload-config option work again

### PR DESCRIPTION
**Pull request by paulidale**
The no-autoload-config option stopped working after 1.1.1.
It should work still.

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Review by mattcaswell — APPROVED** [review]

**Comment by evgenykotkov** [comment]
Hi Paul!

Thanks for taking care of this regression.  As a small side opinion: maybe introducing a helper such as `ossl_ensure_config_loaded()` would make the code easier to maintain?

This helper could be made no-op if `OPENSSL_NO_AUTOLOAD_CONFIG` is defined and would allow to:

- Remove the potentially error-prone `#if !defined … && !defined …` conditions
- Skip having to check for `OPENSSL_NO_AUTOLOAD_CONFIG` everywhere where it's needed now and in the future

**Comment by paulidale** [comment]
@evgenykotkov, I did contemplate adding such a call but while making the changes it didn't seem worthwhile given the other code I could remove.  Some of the other conditional checks looked like being expensive and they can also be avoided by this approach.

Still, such a call might be worth introducing at some point.

**Comment by paulidale** [comment]
Ping @openssl/committers for second review please.

File: test/pbetest.c
Comment (slontis) [inline test/pbetest.c:145]: return 0.. otherwise main.c will run tests and call cleanup again.

Comment (slontis) [reply test/pbetest.c:145]: Ahh i see you have changed the beviour of main.. so that it doesnt print usage for -1.

File: crypto/asn1/a_strnid.c
Comment (slontis) [inline crypto/asn1/a_strnid.c:134]: Another way of doing this would be to put the code into OPENSSL_init_crypto instead..
````
#ifdef OPENSSL_NO_AUTOLOAD_CONFIG
 if (opts == OPENSSL_INIT_LOAD_CONFIG)
      return 1;
#endif

Comment (slontis) [reply crypto/asn1/a_strnid.c:134]: That would be future proof..

Comment (paulidale) [reply crypto/asn1/a_strnid.c:134]: No it wouldn't.  Calling OPENSSL_init_crypto with the OPENSSL_INIT_LOAD_CONFIG flag is perfectly legitimate from an application even with this compile time option.  What mustn't happen is the library ever making such a call.

**Review by slontis — APPROVED** [review]

File: test/pbetest.c
Comment (paulidale) [reply test/pbetest.c:145]: Yeah, printing usage here was useless and I needed a way to signal this back to the framework,

**Comment by paulidale** [comment]
Merged to all three branches.  OpenSSL 3.0 required a manual fix but it was fairly trivial.
Thanks for the reviews.

**Comment by t8m** [comment]
I think this deserves a CHANGES.md entry.

**Pull request closed without merging**
